{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from ete3 import NCBITaxa\n",
    "ncbi = NCBITaxa()\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# load uhgg metadata and blast results\n",
    "bacteria_db_metadata = pd.read_csv(\n",
    "# '/home/carsonjm/resources/uhgg/genome-all_metadata.tsv'\n",
    "str(snakemake.input.bacteria_db_metadata)\n",
    ", sep='\\t')\n",
    "\n",
    "# split lineage column\n",
    "bacteria_db_metadata[['superkingdom', 'phylum', 'class', 'order','family', 'genus', 'species']] = bacteria_db_metadata['Lineage'].str.split(';', expand=True)\n",
    "bacteria_db_metadata['superkingdom'] = bacteria_db_metadata['superkingdom'].str.partition('d__')[2]\n",
    "bacteria_db_metadata['phylum'] = bacteria_db_metadata['phylum'].str.partition('p__')[2]\n",
    "bacteria_db_metadata['class'] = bacteria_db_metadata['class'].str.partition('c__')[2]\n",
    "bacteria_db_metadata['order'] = bacteria_db_metadata['order'].str.partition('o__')[2]\n",
    "bacteria_db_metadata['family'] = bacteria_db_metadata['family'].str.partition('f__')[2]\n",
    "bacteria_db_metadata['genus'] = bacteria_db_metadata['genus'].str.partition('g__')[2]\n",
    "bacteria_db_metadata['superkingdom'] = bacteria_db_metadata['superkingdom'].str.partition('_')[0]\n",
    "bacteria_db_metadata['phylum'] = bacteria_db_metadata['phylum'].str.partition('_')[0]\n",
    "bacteria_db_metadata['class'] = bacteria_db_metadata['class'].str.partition('_')[0]\n",
    "bacteria_db_metadata['order'] = bacteria_db_metadata['order'].str.partition('_')[0]\n",
    "bacteria_db_metadata['family'] = bacteria_db_metadata['family'].str.partition('_')[0]\n",
    "bacteria_db_metadata['genus'] = bacteria_db_metadata['genus'].str.partition('_')[0]\n",
    "\n",
    "\n",
    "# read phist report line-by-line\n",
    "phist = open(str(snakemake.input.phist), 'r')\n",
    "\n",
    "Lines = phist.readlines()\n",
    "\n",
    "# create lists to store data in\n",
    "phage_col = []\n",
    "host_col = []\n",
    "kmer_col = []\n",
    "\n",
    "# if there are multiple common kmers, write out phage, host, and # kmers\n",
    "for line in Lines:\n",
    "    if \"kmer-length:\" in line:\n",
    "        viruses = line.split(',')\n",
    "        viruses.pop(0)\n",
    "        viruses.pop(0)\n",
    "        viruses.pop()\n",
    "    elif \"query-samples\" in line:\n",
    "        continue\n",
    "    elif ':' in line:\n",
    "        bacterial_genome = line.split(',')[0]\n",
    "        hits = line.split(',')[1:]\n",
    "        hits.pop(0)\n",
    "        hits.pop()\n",
    "        for hit in hits:\n",
    "            index, kmers = str(hit).split(':')\n",
    "            if int(kmers) > 1:\n",
    "                phage_col.append(str(viruses[int(index) - 1]))\n",
    "                host_col.append(str(bacterial_genome))\n",
    "                kmer_col.append(int(kmers))\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "phist.close()\n",
    "\n",
    "phist_results = pd.DataFrame()\n",
    "phist_results['phage'] = phage_col\n",
    "phist_results['host'] = host_col\n",
    "phist_results[\"common_kmers\"] = kmer_col\n",
    "\n",
    "# format blast genome so it can be merged with metadata, then merge\n",
    "phist_results['Genome'] = phist_results['host'].str.split('.gff.fasta', expand=True)[0]\n",
    "phist_results_metadata = phist_results.merge(bacteria_db_metadata, on='Genome')\n",
    "\n",
    "# filter to only retain genomes with significant number of common kmers\n",
    "phist_results_blast = phist_results_metadata[phist_results_metadata['common_kmers'] >\n",
    "snakemake.params.min_blast_kmers\n",
    "# 0.5\n",
    "]\n",
    "\n",
    "blast_genomes = pd.DataFrame()\n",
    "blast_genomes['host'] = list(set(phist_results_blast['host']))\n",
    "blast_genomes['host'] = str(snakemake.params.bacteria_db_dir) + \"/\" +  blast_genomes['host'].astype(str)\n",
    "blast_genomes.to_csv(str(snakemake.output.blast_genomes), index=False, header=None)\n",
    "\n",
    "# filter to only retain genomes with significant number of common kmers\n",
    "phist_results_metadata_hq = phist_results_metadata[phist_results_metadata['common_kmers'] >\n",
    "snakemake.params.min_common_kmers\n",
    "# 0.5\n",
    "]\n",
    "\n",
    "merged=phist_results_metadata_hq[['phage', 'host', 'common_kmers', 'superkingdom', 'phylum', 'class', 'order', 'family', 'genus']]\n",
    "\n",
    "merged2 = merged[merged['host'].notnull()]\n",
    "\n",
    "# count total common kmers\n",
    "phist_count = merged2.groupby(by=['phage'], as_index=False).sum()\n",
    "phist_counts = phist_count.loc[:,['phage', 'common_kmers']]\n",
    "phist_counts.rename(columns = {'common_kmers':'total_kmers'}, inplace = True)\n",
    "phist_counts_taxonomy = merged2.merge(phist_counts, on='phage', how='left')\n",
    "\n",
    "# determine if any genomes have > min_agreement agreement at taxonomic level\n",
    "def determine_consensus(taxonomic_rank, phist_table):\n",
    "    taxonomy_hits = phist_table.groupby(['phage', taxonomic_rank], as_index=False\n",
    "            ).agg(phist_taxonomy_hits=('common_kmers' ,'sum'))\n",
    "    taxonomy_hits.rename(columns={'phist_taxonomy_hits':taxonomic_rank + '_kmers'}, inplace=True)\n",
    "    taxonomy_hits_merged = phist_table.merge(taxonomy_hits, on=['phage', taxonomic_rank], how='outer')\n",
    "    taxonomy_hits_merged2 = taxonomy_hits_merged[taxonomy_hits_merged[taxonomic_rank + \"_kmers\"].notnull()]\n",
    "    taxonomy_hits_merged2[taxonomic_rank + '_percent_agreement'] = taxonomy_hits_merged2[taxonomic_rank + '_kmers'].astype(int)/taxonomy_hits_merged2['total_kmers']\n",
    "    taxonomy_hits_consensus = taxonomy_hits_merged2[taxonomy_hits_merged2[taxonomic_rank + '_percent_agreement']*100 >\n",
    "    # 70\n",
    "    snakemake.params.min_agreement\n",
    "    ]\n",
    "    taxonomy_hits_consensus = taxonomy_hits_consensus.groupby('phage', as_index=False).first()\n",
    "\n",
    "    return taxonomy_hits_consensus\n",
    "\n",
    "# determine genus level consensus\n",
    "genus_consensus = determine_consensus('genus', phist_counts_taxonomy)\n",
    "genus_unannotated = phist_counts_taxonomy[~phist_counts_taxonomy['phage'].isin(\n",
    "    genus_consensus['phage'])]\n",
    "\n",
    "\n",
    "# determine family level consensus\n",
    "family_consensus = determine_consensus('family', genus_unannotated)\n",
    "family_consensus['genus'] = 'NA'\n",
    "family_unannotated = genus_unannotated[~genus_unannotated['phage'].isin(\n",
    "    family_consensus['phage'])]\n",
    "\n",
    "# determine order level consensus\n",
    "order_consensus = determine_consensus('order', family_unannotated)\n",
    "order_consensus[['genus', 'family']] = 'NA'\n",
    "order_unannotated = family_unannotated[~family_unannotated['phage'].isin(\n",
    "    order_consensus['phage'])]\n",
    "\n",
    "# determine class level consensus\n",
    "class_consensus = determine_consensus('class', order_unannotated)\n",
    "class_consensus[['genus', 'family', 'order']] = 'NA'\n",
    "class_unannotated = order_unannotated[~order_unannotated['phage'].isin(\n",
    "    class_consensus['phage'])]\n",
    "\n",
    "# determine phylum level consensus\n",
    "phylum_consensus = determine_consensus('phylum', class_unannotated)\n",
    "phylum_consensus[['genus', 'family', 'order', 'class']] = 'NA'\n",
    "phylum_unannotated = class_unannotated[~class_unannotated['phage'].isin(\n",
    "    phylum_consensus['phage'])]\n",
    "\n",
    "# determine superkingdom level consensus\n",
    "superkingdom_consensus = determine_consensus('superkingdom', phylum_unannotated)\n",
    "superkingdom_consensus[['genus', 'family', 'order', 'class', 'phylum']] = 'NA'\n",
    "superkingdom_unannotated = phylum_unannotated[~phylum_unannotated['phage'].isin(\n",
    "    superkingdom_consensus['phage'])]\n",
    "\n",
    "# merge all results\n",
    "final_consensus = pd.concat([genus_consensus, family_consensus, order_consensus,\n",
    "    class_consensus, phylum_consensus, superkingdom_consensus])\n",
    "final_consensus = final_consensus.fillna('NA')\n",
    "final_consensus['taxonomy'] = (final_consensus['superkingdom'] + ';'\n",
    "    + final_consensus['phylum'] + ';' + final_consensus['class'] + ';'\n",
    "    + final_consensus['order'] + ';' + final_consensus['family'] + ';'\n",
    "    + final_consensus['genus'])\n",
    "\n",
    "# format results file and save\n",
    "final_consensus = final_consensus[['phage', 'common_kmers', 'total_kmers',\n",
    "'superkingdom','superkingdom_kmers','superkingdom_percent_agreement','phylum','phylum_kmers','phylum_percent_agreement',\n",
    "'class','class_kmers','class_percent_agreement','order','order_kmers','order_percent_agreement',\n",
    "'family','family_kmers','family_percent_agreement','genus','genus_kmers','genus_percent_agreement', 'taxonomy']]\n",
    "final_consensus = final_consensus.add_prefix('phist_')\n",
    "final_consensus.rename(columns = {'phist_phage': 'viral_genome'}, inplace=True)\n",
    "final_consensus['viral_genome'] = final_consensus['viral_genome'].str.split('.fna', expand=True)[0]\n",
    "final_consensus.to_csv(str(snakemake.output.report), index=False)\n",
    "host_results = final_consensus[['viral_genome', 'phist_taxonomy']]\n",
    "host_results.to_csv(str(snakemake.output.taxonomy), index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": ""
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "25a19fbe0a9132dfb9279d48d161753c6352f8f9478c2e74383d340069b907c3"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
