{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "from operator import itemgetter\n",
    "\n",
    "# parse and save record.id of all present viruses\n",
    "present_viruses = []\n",
    "\n",
    "for record in SeqIO.parse(str(snakemake.input.viruses), \"fasta\"):\n",
    "    present_viruses.append(record.id)\n",
    "\n",
    "present_viruses_set = set(present_viruses)\n",
    "\n",
    "# open metadata and save virus-host connections\n",
    "metadata = pd.read_csv(str(snakemake.input.metadata), sep='\\t')\n",
    "metadata[\"UViG_split\"] = metadata[\"## UViG\"].str.partition(\"|\")[0]\n",
    "metadata[['superkingdom', 'phylum', 'class', 'order','family', 'genus', 'species']] = metadata['Host taxonomy prediction'].str.split(';', expand=True)\n",
    "metadata['phylum'] = metadata['superkingdom'].astype(str) + ';' + metadata['phylum'].astype(str)\n",
    "metadata['class'] = metadata['phylum'].astype(str) + ';' + metadata['class'].astype(str)\n",
    "metadata['order'] = metadata['class'].astype(str) + ';' + metadata['order'].astype(str)\n",
    "metadata['family'] = metadata['order'].astype(str) + ';' + metadata['family'].astype(str)\n",
    "metadata['genus'] = metadata['family'].astype(str) + ';' + metadata[\"genus\"].astype(str)\n",
    "uvigs = list(metadata[\"UViG_split\"])\n",
    "uvig_genus = metadata.set_index('UViG_split').to_dict()['genus']\n",
    "uvig_family = metadata.set_index('UViG_split').to_dict()['family']\n",
    "uvig_order = metadata.set_index('UViG_split').to_dict()['order']\n",
    "uvig_class = metadata.set_index('UViG_split').to_dict()['class']\n",
    "uvig_phylum = metadata.set_index('UViG_split').to_dict()['phylum']\n",
    "uvig_superkingdom = metadata.set_index('UViG_split').to_dict()['superkingdom']\n",
    "\n",
    "# define consensus function\n",
    "def taxa_consensus(node_list, taxa_map):\n",
    "    taxa = []\n",
    "    for node in node_list:\n",
    "        node2 = node.split('|')[0]\n",
    "        if node2 in taxa_map:\n",
    "            taxa.append(taxa_map[node2])\n",
    "    freq = Counter(taxa)\n",
    "    num_hits = len(taxa)\n",
    "    if num_hits > 0:\n",
    "        agreement = max(freq.values())/len(taxa)\n",
    "        if agreement > snakemake.params.min_agreement:\n",
    "            consensus = [key for key, value in freq.items() if value == max(freq.values())]\n",
    "        else:\n",
    "            consensus = \"\"\n",
    "            agreement = 0\n",
    "    else:\n",
    "        agreement = 0\n",
    "        consensus = \"\"\n",
    "    return num_hits, agreement, consensus\n",
    "\n",
    "# iterate through cluster file\n",
    "votu = []\n",
    "genus = []\n",
    "genus_hits = []\n",
    "genus_agreement = []\n",
    "family = []\n",
    "family_hits = []\n",
    "family_agreement = []\n",
    "order = []\n",
    "order_hits = []\n",
    "order_agreement = []\n",
    "classes = []\n",
    "class_hits = []\n",
    "class_agreement = []\n",
    "phylum = []\n",
    "phylum_hits = []\n",
    "phylum_agreement = []\n",
    "superkingdom = []\n",
    "superkingdom_hits = []\n",
    "superkingdom_agreement = []\n",
    "\n",
    "# open clustering results\n",
    "clusters = open(\n",
    "    str(snakemake.input.clusters) , 'r')\n",
    "\n",
    "\n",
    "for line in clusters:\n",
    "    stripped = line.strip()\n",
    "    centroid, nodes = stripped.split('\\t')\n",
    "    if centroid in present_viruses_set:\n",
    "        votu.append(centroid.split('|')[0])\n",
    "        nodes_split = nodes.split(\",\")\n",
    "        hits, agreement, consensus = taxa_consensus(nodes_split, uvig_genus)\n",
    "        genus_hits.append(hits)\n",
    "        genus_agreement.append(agreement)\n",
    "        genus.append(consensus)\n",
    "        hits, agreement, consensus = taxa_consensus(nodes_split, uvig_family)\n",
    "        family_hits.append(hits)\n",
    "        family_agreement.append(agreement)\n",
    "        family.append(consensus)\n",
    "        hits, agreement, consensus = taxa_consensus(nodes_split, uvig_order)\n",
    "        order_hits.append(hits)\n",
    "        order_agreement.append(agreement)\n",
    "        order.append(consensus)\n",
    "        hits, agreement, consensus = taxa_consensus(nodes_split, uvig_class)\n",
    "        class_hits.append(hits)\n",
    "        class_agreement.append(agreement)\n",
    "        classes.append(consensus)\n",
    "        hits, agreement, consensus = taxa_consensus(nodes_split, uvig_phylum)\n",
    "        phylum_hits.append(hits)\n",
    "        phylum_agreement.append(agreement)\n",
    "        phylum.append(consensus)\n",
    "        hits, agreement, consensus = taxa_consensus(nodes_split, uvig_superkingdom)\n",
    "        superkingdom_hits.append(hits)\n",
    "        superkingdom_agreement.append(agreement)\n",
    "        superkingdom.append(consensus)\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "\n",
    "dict = {'genome':votu,\n",
    "        'genus':genus, 'genus_hits':genus_hits, 'genus_agreement':genus_agreement,\n",
    "        'family':family, 'family_hits':family_hits, 'family_agreement':family_agreement,\n",
    "        'order':order, 'order_hits':order_hits, 'order_agreement':order_agreement,\n",
    "        'classes':classes, 'class_hits':class_hits, 'class_agreement':class_agreement,\n",
    "        'phylum':phylum, 'phylum_hits':phylum_hits, 'phylum_agreement':phylum_agreement,\n",
    "        'superkingdom':superkingdom, 'superkingdom_hits':superkingdom_hits, 'superkingdom_agreement':superkingdom_agreement}\n",
    "\n",
    "votu_consensus = pd.DataFrame(dict)\n",
    "\n",
    "votu_consensus = votu_consensus.add_prefix('votu_')\n",
    "votu_consensus.rename(columns={\"votu_genome\":\"viral_genome\"}, inplace=True)\n",
    "votu_consensus.to_csv(str(snakemake.output.report), index=False)\n",
    "\n",
    "votu_consensus['votu_consensus_taxonomy'] = ''\n",
    "votu_consensus['votu_consensus_taxonomy'] = votu_consensus.apply(lambda x: x.votu_superkingdom if x.votu_superkingdom_agreement > snakemake.params.min_agreement else x.votu_consensus_taxonomy, axis=1)\n",
    "votu_consensus['votu_consensus_taxonomy'] = votu_consensus.apply(lambda x: x.votu_phylum if x.votu_phylum_agreement > snakemake.params.min_agreement else x.votu_consensus_taxonomy, axis=1)\n",
    "votu_consensus['votu_consensus_taxonomy'] = votu_consensus.apply(lambda x: x.votu_classes if x.votu_class_agreement > snakemake.params.min_agreement else x.votu_consensus_taxonomy, axis=1)\n",
    "votu_consensus['votu_consensus_taxonomy'] = votu_consensus.apply(lambda x: x.votu_order if x.votu_order_agreement > snakemake.params.min_agreement else x.votu_consensus_taxonomy, axis=1)\n",
    "votu_consensus['votu_consensus_taxonomy'] = votu_consensus.apply(lambda x: x.votu_family if x.votu_family_agreement > snakemake.params.min_agreement else x.votu_consensus_taxonomy, axis=1)\n",
    "votu_consensus['votu_consensus_taxonomy'] = votu_consensus.apply(lambda x: x.votu_genus if x.votu_genus_agreement > snakemake.params.min_agreement else x.votu_consensus_taxonomy, axis=1)\n",
    "\n",
    "votu_consensus[['votu_genome', 'votu_consensus_taxonomy']].to_csv(str(snakemake.output.taxonomy), index=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.7"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "f9f85f796d01129d0dd105a088854619f454435301f6ffec2fea96ecbd9be4ac"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
